Overhaul of filters on integrations page

kknopper opened the pull request:
## Proposed change
A rework of #24500 / #18512 to add proper filtration to the integrations page. Also addresses #9944

[Netlify preview / demo](https://deploy-preview-24642--home-assistant-docs.netlify.app/integrations/)

used @ludeeus 's notes from here as a guideline: [#16996 (comment)](https://github.com/home-assistant/home-assistant.io/issues/16996#issuecomment-880515679)

- [x] Better styling for filters. Thanks @ludeeus for the starting point
- [x] added filtration of IoT Classes, and Quality Scale 
- [x] all filtration items have been moved into `<select>` dropdowns. (Category, Version, IoT, and Quality)
- [x] It's now possible to combine filters (Search, IoT, Version, Category, and Quality)
- [x] an easy way to reset filters (buttons get added to the dom that can be clicked to remove filters)
- [x] complete rework of the `applyFilter()` function that allows for combining filters
- [x] update is backwards compatible with old `#` urls
- [x] on page load, if no filters are present in the url, the category is set to `Featured` like the production site 
- [x] integrations can now be searched for by their iot class and quality scale. You can also search for Featured integrations

Completely reworked the javascript for the integrations page. 
I opted to use the [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) browser API, over the hash method, as this api makes it easy to add and remove querys/filters to the url.

Created a `buildQueryFromURL()` function to build a query object of active filters any time the url gets updated. This object is then compared against the integrations array for filtration. Added more helper functions as well.

Old `#` Urls still work as expected. Also allowed for new hash urls for IoT and Quality which look like so:
`.../integrations/#iot_class/local-polling` and `.../integrations/#quality_scale/internal`

If a `#` url is present, any change to a filter will convert the URL to use the URLSearchParams API format instead.

On page load, or any change to the url, the current active filters (select dropdowns / search bar) will get updated to match the current active filters

Filter removal buttons are dynmaically added to the dom as well, and also update to match current active filters

Filters can be removed by clicking these buttons, or setting a filter dropdown to its initial state

### Caveats:
- URLSearchParams is not supported by IE
- this update contains a lot of ES6 javascript which is not supported by IE
- if IE needs to be supported, there are [polyfills](https://github.com/jerrybendy/url-search-params-polyfill) available for the URLSearchParams api

This is a pretty large change, because of this, I tried to make my code as readable as possible with plenty of comments.
Also left console log's present for reviewing purposes.  

## Type of change
<!--
    What types of changes does your PR introduce to our documentation/website?
    Put an `x` in the boxes that apply. You can also fill these out after
    creating the PR.
-->

- [x] Spelling, grammar or other readability improvements (`current` branch).
- [ ] Adjusted missing or incorrect information in the current documentation (`current` branch).
- [ ] Added documentation for a new integration I'm adding to Home Assistant (`next` branch).
  - [ ] I've opened up a PR to add logo's and icons in [Brands repository](https://github.com/home-assistant/brands).
- [ ] Added documentation for a new feature I'm adding to Home Assistant (`next` branch).
- [ ] Removed stale or deprecated documentation.

## Additional information
<!--
    Details are important, and help maintainers processing your PR.
    Please be sure to fill out additional details, if applicable.
-->

- Link to parent pull request in the codebase: 
- Link to parent pull request in the Brands repository: 
- This PR fixes or closes issue: fixes #9944

## Checklist
<!--
    Put an `x` in the boxes that apply. You can also fill these out after
    creating the PR. If you're unsure about any of them, don't hesitate to ask.
    We're here to help! This is simply a reminder of what we are going to look
    for before merging your code.
-->

- [x] This PR uses the correct branch, based on one of the following:
  - I made a change to the existing documentation and used the `current` branch.
  - I made a change that is related to an upcoming version of Home Assistant and used the `next` branch.
- [x] The documentation follows the Home Assistant documentation [standards].

[standards]: https://developers.home-assistant.io/docs/documenting/standards

Commit:
overhaul of filters on integrations page

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

ludeeus commented on source/integrations/index.html:
```diff
@@ -23,43 +23,87 @@
+  </div>
+  <div class="grid__item three-sixths lap-one-whole palm-one-whole">
+      <div class="integration_filter">
+        <select name="category" data-id="cat">
```
Why not use `ha_category` here (and everywhere else)?

kknopper replied:
added a new `ha_category` class for the 4 select drop downs

ludeeus commented on source/integrations/index.html:
```diff
@@ -23,43 +23,87 @@
+      </form>
+    </div>
+  </div>
+  <div class="grid__item three-sixths lap-one-whole palm-one-whole">
```
I think `one-quarter` looks better:
![image](https://user-images.githubusercontent.com/15093472/203175683-d0bae818-199d-4841-8657-d734a64e3c15.png)

`three-sixths`:
![image](https://user-images.githubusercontent.com/15093472/203175665-4200c2e8-1dfc-4e8f-aa6b-357a59d99558.png)

kknopper replied:
updated to use `one-quarter` class for select dropdown width

ludeeus commented on source/integrations/index.html:
```diff
@@ -23,43 +23,87 @@
+      <select name="iot_classes" data-id="iot_class">
+        <option value="#">IoT Class</option>
+        {%- for iot_class in iot_classes -%}
+          {%- assign iot_class_count = components | where: 'ha_iot_class', iot_class | size -%}
```
Counts do not consider active filters.
I wonder if we should show it at all.

ludeeus commented on source/integrations/index.html:
```diff
@@ -171,99 +318,241 @@
+        for (key in queryFilter) {
+          if (document.querySelector(`.component-search .active-filter#${key}`) === null) {
+            let button = document.createElement('button');
+            button.innerHTML = `<span>${getButtonText(key)}<i class="icon-plus-sign"></i></span>`;
```
this should include the filtered value as well, like `version: 2022.11.0` and `Category: featured`.

kknopper replied:
updated the `updateFilterButtons` function. Now there is another check to test if the filter object values have changed, and it updates the button text accordingly. Also switched to use the data attribute over id

ludeeus commented on source/integrations/index.html:
```diff
@@ -103,60 +147,163 @@
+
+      if (hash.indexOf('#search/') === 0 || queryParams.has('search')) {
+        let query = getQueryValue('search');
+        queryFilter.search = query;
```
```suggestion
        queryFilter.search = query.toLowerCase();
```

ludeeus commented on source/integrations/index.html:
```diff
@@ -171,99 +318,241 @@
+    }
+
+    // update all filter buttons to match current obj filter
+    function updateFilterButtons(queryFilter) {
```
Lets reserve the space these take up so the UI does not jump when adding/removing

kknopper replied:
added height to the container holding the filter buttons, so the ui doesnt jump anymore

kknopper replied:
@ludeeus not sure if there's a better way we wanna handle this for mobile. The filter buttons can stack, and create a lot more space. Adjusted my initial fix to use `min-height`, so it grows for mobile.

![Screen Shot 2022-11-30 at 9 13 43 AM](https://user-images.githubusercontent.com/6854383/204836726-cc25e199-9209-4135-a53f-c9d5f3b3e911.png)

ludeeus commented on sass/custom/_component_page.scss:
```diff
@@ -12,6 +12,35 @@
+    margin-bottom: 10px;
+    text-align: center;
+  
+    &[data-id="iot_class"], &[data-id="quality_scale"] {
+    margin-bottom: 15px;
```
With a single row this can be removed

kknopper replied:
removed

ludeeus commented on sass/custom/_component_page.scss:
```diff
@@ -12,6 +12,35 @@
+
+  .integration_filter select {
+    width: 100%;
+    margin-bottom: 10px;
```
```suggestion
    margin-bottom: 15px;
```

With a single row this can be 15 instead, 16 maybe?

kknopper replied:
updated to 15px

ludeeus commented on sass/custom/_component_page.scss:
```diff
@@ -12,6 +12,35 @@
+  }
+  
+  .active-filter {
+    margin: 10px 5px 0px 0;
```
```suggestion
    margin: 10px 5px 0px 0px;
```

kknopper replied:
cleaned this up

ludeeus commented on source/integrations/index.html:
```diff
@@ -103,60 +147,163 @@
+        $searchFilter.val(null);
+      }
+      
+      if (hash.indexOf('#version/') === 0 || queryParams.has('version')) {
```
```suggestion
      if (hash.startsWith('#version/') || queryParams.has('version')) {
```
The same goes for the rest of these

kknopper replied:
updated all instances of `indexOf` to use `startsWith` method

ludeeus commented on source/integrations/index.html:
```diff
@@ -103,60 +147,163 @@
+    // get query value pair that matches provided id from url. // for buildQueryFromURL()
+    function getQueryValue(id) {
+      let index = id.length + 2; // id + '#' and '/' => '#search/'
+      if (id === 'cat') index = 1;
```
```suggestion
      if (id === 'cat') {
        index = 1;
      }
```

kknopper replied:
cleaned this up

ludeeus commented on source/integrations/index.html:
```diff
@@ -171,99 +318,241 @@
+          text = 'IoT';
+          break;
+        case 'quality_scale':
+          text = 'Quality';
```
```suggestion
          text = 'Quality scale';
```

kknopper replied:
updated filtered button name

ludeeus commented on source/integrations/index.html:
```diff
@@ -171,99 +318,241 @@
+          text = 'Category';
+          break;
+        case 'iot_class':
+          text = 'IoT';
```
```suggestion
          text = 'IoT class';
```

kknopper replied:
updated filter button names

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
addressing PR review feedback

kknopper commented:
@ludeeus addressed PR feedback. Currently I'm leaving in the counts, as this has been useful for testing. 

When you think this PR is review ready, I can add an additional commit to remove all filter counts, along with all the current js console logs.

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
min-height fix for filter button ui

ludeeus commented:
> When you think this PR is review ready, I can add an additional commit to remove all filter counts, along with all the current js console logs.

Please do that.

Commit:
removing console logs

Commit:
removing counts from select dropdowns

Commit:
Other dropdown select fix

Commit:
removing hyphens from filter buttons

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
capitalizing quality scale to match other dropdowns

kknopper commented:
@ludeeus I think this PR is in a pretty good place. 

The only thing i would like to change would be to properly order the dropdown of the Quality Scale select element. Was trying to do this in the liquid syntax, but I couldn't find an elegant solution. I think the simplest way would be to manually declare an array variable, instead of mapping it from the integrations. This doesn't make it dynamic though. Let me know your thoughts on this, along with any other feedback you might have. Would love to get this merged soon!

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

frenck commented:
I think my main grind with this approach is:

![CleanShot 2023-01-31 at 09 39 03](https://user-images.githubusercontent.com/195327/215710241-d8b18466-c8e3-4809-a766-79471d8a0084.png)

The "Featured" is set by default, which is OK, but is also hides the "categories". When getting to this page now the first time, it is not clear where all the categories are...

../Frenck

kknopper commented:
@frenck We could add custom dropdown titles above the select element to make this more clear, similar to the version select thats on the integrations page now:

![Screen Shot 2023-02-03 at 11 07 36 AM](https://user-images.githubusercontent.com/6854383/216664339-6c66ca60-2480-4b58-be88-50d75b9f26ab.png)

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
adding check for dropdowns, better styling for dropdowns, adjusting all selector

kknopper commented:
@frenck @ludeeus 

- Added a title for each of the dropdown selectors
- moved selectors above search bar
- Cleaned up styling for dropdown selectors - made sure styles were compatible cross browser
  - added a wrapper div around dropdowns to make sure widths matched across major browsers (Chrome / FF / Safari)
- Now the default for every selector will be `All` for when no filter is applied 
- adjusted how the `All` filter works for the Categories dropdown
  - old `/integrations/#all` url still works as expected
- Added a check for the dropdowns - now if a user changes the url to a filter that is not present, the dropdown will only change if a matching option is available

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
fixing select dropdown color on mobile

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
fixing a bug with the filter removal buttons

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper commented:
@frenck @ludeeus If you guys have time to take another look at this PR, that would be awesome!

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
adjusted search input text so that it stays consistent with url and url pushState

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul
